modules/serverless-invoking-quarkus-functions.adoc

```diff
@@ -0,0 +1,157 @@
+----
+public class Functions {
+    @Funq
+    public void processPurchase(Purchase purchase) {
+        // process the purchase
+    }
+}
+----
+
+The corresponding `Purchase` JavaBean class that contains the purchase data looks as follows:
+
+.Example class
+[source,java]
+----
+public class Purchase {
+    private long customerId;
+    private long productId;
+    // getters and setters
+}
+----
+
+[id="serverless-invoking-quarkus-functions-examples_{context}"]
+== Invocation examples
+
+The following example code defines three functions named `withBeans`, `withCloudEvent`, and `withBinary`;
+
+.Example
+[source,java]
+----
+import io.quarkus.funqy.Funq;
+import io.quarkus.funqy.knative.events.CloudEvent;
+
+public class Input {
+    private String message;
+
+    // getters and setters
+}
+
+public class Output {
+    private String message;
+
+    // getters and setters
+}
+
+public class Functions {
+    @Funq
+    public Output withBeans(Input in) {
+        // function body
+    }
+
+    @Funq
+    public CloudEvent<Output> withCloudEvent(CloudEvent<Input> in) {
+        // function body
+    }
+
+    @Funq
+    void withBinary(byte[] in) {
+        // function body
+    }
+}
+----
+
+The `withBeans` function of the `Functions` class can be invoked by:
+
+* An HTTP POST request with JSON body:
++
+[source,terminal]
+----
+$ curl "http://localhost:8080/" -X POST \
+    -H "Content-Type: application/json" \
+    -d '{"message": "Hello there."}'
+----
+* An HTTP GET request with query parameters:
++
+[source,terminal]
+----
+$ curl "http://localhost:8080?message=Hello%20there." -X GET
+----
+* A `CloudEvent` object in binary encoding:
++
+[source,terminal]
+----
+curl "http://localhost:8080/" -X POST \
+  -H "Content-Type: application/json" \
+  -H "Ce-SpecVersion: 1.0" \
+  -H "Ce-Type: my-type" \
+  -H "Ce-Source: cURL" \
+  -H "Ce-Id: 42" \
+  -d '{"message": "Hello there."}'
+----
+* A `CloudEvent` object in structured encoding:
++
+[source,terminal]
+----
+$ curl http://localhost:8080/ \
+    -H "Content-Type: application/cloudevents+json" \
+    -d '{ "data": {"message":"Hello there."},
+          "datacontenttype": "application/json",
+          "id": "42",
+          "source": "curl",
+          "type": "my-type",
+          "specversion": "1.0"}'
+----
+
+The `withCloudEvent` function of the `Functions` class can be invoked by using a `CloudEvent` object, similarly to the `withBeans` function. However, unlike `withBeans`, `withCloudEvent` cannot be invoked with a plain HTTP request.
+
+The `withBinary` function of the `Functions` class can be invoked by:
+
+* `CloudEvent` in binary encoding:
++
+[source]
+----
+curl "http://localhost:8080/" -X POST \
+  -H "Content-Type: application/octet-stream" \
+  -H "Ce-SpecVersion: 1.0"\
+  -H "Ce-Type: my-type" \
+  -H "Ce-Source: cURL" \
+  -H "Ce-Id: 42" \
+  --data-binary '@img.jpg'
+----
+* `CloudEvent` in structured encoding:
++
+[source]
+----
+curl http://localhost:8080/ \
```

Comment:
I'm not sure about this one. It didn't work for me

Reply:
@lance any ideas?

Reply:
Note to self: this command is also missing a `$` at the start of the line, will fix later when I have confirmation about the QE comment.

Reply:
`$(base64 img.jpg)` should be `$(base64 --wrap=0 img.jpg)` my bad

Reply:
the base64 util wraps lines at some point, I tested it with little tiny files where it wasn't an issue